_stbt/keyboard.py

```diff
@@ -7,74 +7,68 @@
 from __future__ import absolute_import
 from builtins import *  # pylint:disable=redefined-builtin,unused-wildcard-import,wildcard-import,wrong-import-order
 
+import re
 import time
-from collections import namedtuple
 from logging import getLogger
 
 import networkx as nx
 import numpy
+from attr import attrs, attrib
+from _stbt.grid import Grid
 from _stbt.imgutils import load_image
-from _stbt.utils import text_type
+from _stbt.types import Region
+from _stbt.utils import basestring, text_type
 
 
 log = getLogger("stbt.keyboard")
 
 
+@attrs(frozen=True)
+class Key(object):
+    """A node in the directed graph, representing a key on the keyboard.
+
+    This is an immutable object so that it is hashable (it must be hashable
+    so that we can use it as a node of networkx graphs).
+
+    The user can specify any of the attributes (but at least ``name``). See
+    `Keyboard.add_key`.
+
+    This is an implementation detail, not part of the public API. But we do
+    return these as opaque objects from `Keyboard.add_key`.
+    """
+    name = attrib(default=None, type=text_type)
+    text = attrib(default=None, type=text_type)
+    region = attrib(default=None, type=Region)
+    mode = attrib(default=None, type=text_type)
+
+
+SYMMETRICAL_KEYS = {
+    "KEY_DOWN": "KEY_UP",
+    "KEY_UP": "KEY_DOWN",
+    "KEY_LEFT": "KEY_RIGHT",
+    "KEY_RIGHT": "KEY_LEFT",
+}
+
+
 class Keyboard(object):
```

Comment:
For the KEY_GREEN example, specifically, I think you'd handle it explicitly in your page object's `enter_text` method.

To have the graph handle it, you'd need an edge from every key to itself labelled with `key="KEY_GREEN"` -- but where do you specify the text to type (space)? Is it associated with the edge itself, rather than the node? How would that work?

Reply:
I've been thinking about this and I think we would add an additional node for these transitions.  So it would look like:

![image](https://user-images.githubusercontent.com/494659/93872680-6b2d2180-fcc8-11ea-9673-bbbc3e626155.png)

You'd navigate to the nearest side-effect node when trying to type a letter, and then when on one take the noop path out automatically.